components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,297 @@
+    for (auto const& segment_id : segment_ids) {
+        files.emplace_back(FileInfo{segment_id, offset});
+        get_file_size_and_update_offset(segment_dir_path / segment_id, offset);
+    }
+
+    // Add sentinel indicating total size of all files.
+    files.emplace_back(FileInfo{"", offset});
+
+    // Decompression of large single-file archives will consume excessive memory since
+    // single-file archives are not split.
+    if (offset > cFileSizeWarningThreshold) {
+        SPDLOG_WARN(
+                "Single file archive size exceeded {}. "
+                "The single-file archive format is not intended for large archives, "
+                " consider using multi-file archive format instead.",
+                cFileSizeWarningThreshold
+        );
+    }
+
+    return files;
+}
+
+auto pack_single_file_archive_metadata(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::vector<FileInfo> const& file_infos,
+        size_t num_segments
+) -> std::stringstream {
+    MultiFileArchiveMetadata archive_metadata{
+            .archive_format_version = multi_file_archive_metadata.get_archive_format_version(),
+            .variable_encoding_methods_version
+            = multi_file_archive_metadata.get_variable_encoding_methods_version(),
+            .variables_schema_version = multi_file_archive_metadata.get_variables_schema_version(),
+            .compression_type = multi_file_archive_metadata.get_compression_type(),
+            .creator_id = multi_file_archive_metadata.get_creator_id(),
+            .begin_timestamp = multi_file_archive_metadata.get_begin_timestamp(),
+            .end_timestamp = multi_file_archive_metadata.get_end_timestamp(),
+            .uncompressed_size = multi_file_archive_metadata.get_uncompressed_size_bytes(),
+            .compressed_size = multi_file_archive_metadata.get_compressed_size_bytes(),
+    };
+
+    SingleFileArchiveMetadata single_file_archive{
+            .archive_files = file_infos,
+            .archive_metadata = archive_metadata,
+            .num_segments = num_segments,
+    };
+
+    std::stringstream buf;
+    msgpack::pack(buf, single_file_archive);
+
+    return buf;
+}
+
+auto write_archive_header(FileWriter& archive_writer, size_t packed_metadata_size) -> void {
+    SingleFileArchiveHeader header{
+            .magic{},
+            .version = cArchiveVersion,
+            .metadata_size = packed_metadata_size,
+            .unused{}
+    };
+    std::memcpy(&header.magic, cUnstructuredSfaMagicNumber.data(), sizeof(header.magic));
+    archive_writer.write(reinterpret_cast<char const*>(&header), sizeof(header));
+}
+
+auto write_archive_metadata(FileWriter& archive_writer, std::stringstream const& packed_metadata)
+        -> void {
+    archive_writer.write(packed_metadata.str().data(), packed_metadata.str().size());
+}
+
+auto write_archive_file(std::filesystem::path const& file_path, FileWriter& archive_writer)
```

Comment:
I don't think this is a function specfic for "archive_file". Any chance we can put it into one of the utils.cpp?

Reply:
I don't entirely agree with you, but sure let's leave the code as it is.